envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoySnapshotFactory.kt

```diff
@@ -54,54 +53,94 @@ internal class EnvoySnapshotFactory(
                 it.servicesState.serviceNameToInstances.values
             }.groupBy {
                 it.serviceName
-            }.map { (serviceName, instances) ->
+            }.mapValues { (serviceName, instances) ->
                 toClusterConfiguration(instances, serviceName)
             }
         } else {
-            serviceNames.map { ClusterConfiguration(serviceName = it, http2Enabled = false) }
+            return servicesStates
+                .flatMap { it.servicesState.serviceNames() }
+                .distinct()
+                .associateWith { ClusterConfiguration(serviceName = it, http2Enabled = Http2Status.DISABLED) }
         }
+    }
 
-        val clusters = clustersFactory.getClustersForServices(clusterConfigurations, communicationMode)
+    fun modifyClustersUsingPreviousState(
+        previous: Map<String, ClusterConfiguration>,
+        current: Map<String, ClusterConfiguration>
+    ): Map<String, ClusterConfiguration> {
 
-        val endpoints: List<ClusterLoadAssignment> = createLoadAssignment(servicesStates)
-        val routes = listOf(
-            egressRoutesFactory.createEgressRouteConfig("", serviceNames.map {
-                RouteSpecification(
-                    clusterName = it,
-                    routeDomain = it,
-                    settings = defaultDependencySettings
-                )
-            }, false),
-            ingressRoutesFactory.createSecuredIngressRouteConfig(ProxySettings())
-        )
+        val anyRemoved = if (properties.egress.neverRemoveClusters) {
+            previous.keys.any { it !in current }
+        } else {
+            false
+        }
+        val anyUnknownHttp2 = current.any { it.value.http2Enabled == Http2Status.UNKNOWN }
 
-        val version = snapshotsVersions.version(
-            group = AllServicesGroup(communicationMode, listenersConfig = null),
-            clusters = clusters,
-            endpoints = endpoints
-        )
+        if (!anyRemoved && !anyUnknownHttp2) {
+            // fast path (most frequent)
+            return current
+        }
 
-        val snapshot = createSnapshot(
-            clusters = clusters,
-            clustersVersion = version.clusters,
-            endpoints = endpoints,
-            endpointsVersions = version.endpoints,
-            routes = routes,
-            routesVersion = RoutesVersion(version.clusters.value)
-        )
-        sample.stop(meterRegistry.timer("snapshot-factory.new-snapshot.time"))
+        val fixedClusters = when (anyUnknownHttp2) {
+            true -> fixHttp2StatusUsingPreviousState(previous, current)
+            false -> current
+        }
+        return when (anyRemoved) {
+            true -> {
+                val removedClusters = previous - current.keys
+                fixedClusters + removedClusters
+            }
+            false -> fixedClusters
+        }
+    }
 
-        return snapshot
+    private fun fixHttp2StatusUsingPreviousState(
```

Comment:
WDYT about renaming this method to: `setHttp2StatusUsingPreviousState`? fix looks not so good

Reply:
done